### PR TITLE
fix(openclaw): bind credentials to trusted origin

### PR DESCRIPTION
**Pull request by SweetSophia**
## Functional Summary

This change closes a credential-misrouting gap in the OpenClaw Noosphere memory plugin by making the remote endpoint authority independent from mutable plugin configuration. Previously, an actor who could change `config.baseUrl` could redirect authenticated requests carrying an API key read from protected secret/environment sources. The fix pins non-loopback endpoints to a single trusted HTTPS origin that is read exclusively from the process environment.

### Origin Pinning and Config Validation (`openclaw-noosphere-memory/src/config.ts`)

- Introduces `NoosphereConfigError` as the single, explicit failure signal for configuration problems.
- Adds `assertTrustedRemoteOrigin`, which checks non-loopback base URLs against a `OPENCLAW_NOOSPHERE_TRUSTED_ORIGIN` value (with `NOOSPHERE_TRUSTED_ORIGIN` as a compatibility fallback). The trusted origin is read **only** from the process environment—never from plugin config—so mutable configuration cannot self-authorize a destination.
- `normalizeBaseUrl` now throws instead of silently falling back to `localhost` for malformed URLs, non-HTTP(S) schemes, embedded credentials, query strings/fragments, internal/loopback/non-public hosts, and non-loopback HTTP. Query/fragment rejection is new and prevents URL-parse-based bypasses.
- `normalizeTrustedOrigin` enforces HTTPS-only, public, pathless origins without credentials, query, or fragment. Loopback and private hosts are also rejected as trusted origins.

### Address Classification Improvements

- `isLoopbackHost` now strips trailing dots, parses canonical IPv4, and recognizes IPv4-mapped IPv6 (`::ffff:127.x.x.x`) as loopback. Stops trusting the previous naïve `startsWith("127.")` check, which allowed `127.attacker.example` to slip through as loopback.
- `isBlockedInternalHost` uses a unified `parseIpv4MappedIpv6` helper covering both dotted-decimal and hexadecimal IPv4-mapped forms, ensuring internal IPv4-mapped IPv6 hosts are rejected even before the public-host check.
- `isPrivateOrReservedIpv6` now parses hextets and correctly classifies unique-local (`fc00::/7`), link-local (`fe80::/10`), deprecated site-local (`fec0::/10`, `feff::/8`), multicast (`ff00::/8`), IPv4-compatible (`::/96`), discard-only (`100::/64`), documentation (`2001:db8::/32`), and the unspecified address. The previous string-prefix logic missed several of these.

### HTTP Client Hardening (`openclaw-noosphere-memory/src/client.ts`)

- Sets `redirect: "error"` on every authenticated fetch so 3xx responses cannot carry the API-key-bearing POST body to a different origin. Documented intent: this complements the origin pin so even a misconfigured or compromised upstream cannot redirect credentials cross-origin.

### Tests (`src/__tests__/memory/openclaw-plugin-auto-recall.test.ts`, `noosphere-plugin-env.test.ts`)

- Negative coverage: malformed URLs, internal/private/reserved IPv4 and IPv6 literals, IPv4-mapped private hosts, deprecated site-local/discard prefixes, non-loopback HTTP, mismatched or malformed trusted origins, embedded credentials in the pin, and path/query-bearing pins all throw `NoosphereConfigError`.
- Adversarial coverage: an attacker-controlled `baseUrl` with a `trustedOrigins` field in plugin config fails configuration before any network call (`fetchCalls === 0`), proving plugin config cannot authorize itself.
- Positive coverage: loopback URLs (`localhost`, `127.0.0.1`, `::1`, `localhost.`, IPv4-mapped loopback) are accepted without a pin; remote URLs matching the pinned origin resolve correctly.
- Behavioral coverage: the client passes `redirect: "error"` to `fetch` for authenticated requests, verified on the save path.
- Existing CLI/status/doctor tests were updated from a non-loopback `https://noosphere.local` fixture to `http://127.0.0.1:6578` to keep them loopback and pin-free.

### Documentation

- `openclaw-noosphere-memory/README.md` adds a "Remote origin binding" section describing the variable, its format, why it must live with the API key, and the security boundary.
- `docs/OPENCLAW-OFFICIAL-PLUGIN-SETUP.md` updates the `baseUrl` table row and the security-guidance list to reflect the new requirement and the redirect rejection behavior.

## Operator Impact

- Loopback deployments: no change; no pin required.
- Remote deployments: must set `OPENCLAW_NOOSPHERE_TRUSTED_ORIGIN=https://host[:port]` (scheme/host/port only) in the protected gateway environment before upgrading. Missing, malformed, or mismatched pins now fail initialization explicitly rather than silently redirecting to localhost.

---

## Summary

This PR strengthens the OpenClaw plugin's origin binding by adding tests that verify the credential-pin check properly handles tricky URL inputs and produces canonical hostnames.

## Changes

- **New negative test cases for untrusted origins**: Added URL variants that must be rejected, including URLs with query strings (`?x=1`), fragments (`#frag`), embedded credentials (`user:pass@`), and IPv6 zone identifiers (`fe80::1%eth0`). All of these must be rejected even if the hostname matches the trusted pin.
- **IDN/punycode normalization**: Added a test confirming that an IDN baseUrl host (`bücher.example.test`) is correctly matched against a punycode trusted origin (`xn--bcher-kva.example.test`) and canonicalized to the punycode form.
- **Default HTTPS port handling**: Added a test ensuring `https://host:443` is recognized as the same origin as `https://host` and normalized to drop the default port.
- **IPv4 literal canonicalization**: Added tests verifying dword (`2130706433`), octal-prefixed (`0177.0.0.1`), and short-form (`127.1`) IPv4 loopback representations are all canonicalized to `127.0.0.1` before being checked against the trusted origin.

Together these tests guard against origin-confusion attacks where attackers use alternative URL encodings, embedded credentials, or non-standard IP representations to bypass the trusted-origin credential pin.

---

## Summary

Tightens credential handling for the OpenClaw Noosphere plugin by:

1. **Excluding trusted-origin variables from provider auth metadata** – Verifies the plugin manifest exposes only the two intended `NOOSPHERE_*` API key variables and does not leak any `OPENCLAW_NOOSPHERE_TRUSTED_ORIGIN` setting into provider auth metadata.
2. **Re-enforcing credential/query/fragment rejection regardless of pinning** – Credentials (`user:pass@`), query strings, and URL fragments are now rejected with `NoosphereConfigError` even when `OPENCLAW_NOOSPHERE_TRUSTED_ORIGIN` is set to a valid pinned origin. Previously these three cases only failed in the unpinned path; they are now blocked universally.
3. **Treating scoped IPv6 zone identifiers as invalid URL syntax** – URLs like `https://[fe80::1%25eth0]` are now rejected with the generic “valid HTTP or HTTPS URL” message, classified as invalid syntax rather than as a private-host violation.

### Test changes
- Adds a manifest-assertion test that pins the OpenClaw provider's `envVars` to `["OPENCLAW_NOOSPHERE_API_KEY", "NOOSPHERE_API_KEY"]`, locking out any future accidental inclusion of `OPENCLAW_NOOSPHERE_TRUSTED_ORIGIN`.
- Splits the URL rejection suite: credentials, query strings, and fragments now have their own test that runs against an explicitly pinned trusted origin, while scoped IPv6 zones are folded into a dedicated “invalid URL syntax” case.

### Functional impact
Credentials, query parameters, and fragments embedded in the configured Noosphere base URL can never be used by the plugin, regardless of whether `OPENCLAW_NOOSPHERE_TRUSTED_ORIGIN` is set, closing a path where a pinned-but-malformed URL could otherwise have reached the network layer.

---

**Summary**

This change tightens validation of the Noosphere base URL so that empty query strings and fragments (e.g., a trailing `?` or `#`) are rejected as well, not just non-empty ones. The previous check relied on the parsed `URL` object's `search`/`hash` properties, which treat an empty `?` or `#` as an empty string and therefore passed validation. Switching the check to look directly for `?` or `#` characters in the trimmed input catches these edge cases.

**Why**

Allowing a base URL like `https://noosphere.example.test?` or `https://noosphere.example.test#` is inconsistent with the existing rule that no query strings or fragments are permitted, and the credential-rejection logic relies on the same kind of strict input handling. This keeps the configuration surface predictable and prevents malformed origins from slipping through validation.

**Changes**

- `openclaw-noosphere-memory/src/config.ts`: Replaced the `url.search || url.hash` check with a direct substring check for `?` and `#` in the trimmed input so that empty query/fragment markers are also rejected.
- `src/__tests__/memory/openclaw-plugin-auto-recall.test.ts`: Added test cases covering `?`-only and `#`-only base URLs to lock in the new behavior.

## Summary by Sourcery

Harden OpenClaw Noosphere credential routing by enforcing protected remote-origin binding and strict endpoint validation.

Bug Fixes:
- Bind remote OpenClaw Noosphere credentials to a single protected HTTPS origin and reject mismatched or missing origin pins instead of falling back to localhost.
- Prevent authenticated request bodies from following redirects to another origin.
- Reject malformed, credential-bearing, query-bearing, fragment-bearing, non-public, and non-loopback HTTP endpoints with explicit configuration errors.

Enhancements:
- Improve canonical handling and classification of loopback, IPv4-mapped, private, reserved, and special IPv6 addresses.
- Keep trusted-origin settings out of provider authentication metadata.

Documentation:
- Document remote origin binding requirements, protected environment variables, initialization failures, and redirect rejection behavior.

Tests:
- Expand coverage for origin confusion, URL canonicalization, internal address rejection, environment isolation, and redirect credential protection.

## Summary by CodeRabbit

- **Security Enhancements**
  - Added strict validation for service URLs, blocking unsafe, malformed, or untrusted remote endpoints.
  - Remote HTTPS connections now require a matching trusted-origin setting.
  - Credentialed requests reject HTTP redirects instead of following them.
  - Improved protection against private, reserved, and special network addresses.

- **Documentation**
  - Added setup guidance for local and remote deployments, trusted origins, environment variables, and related configuration errors.

### CODE REVIEW

File: openclaw-noosphere-memory/src/config.ts [hunk 1]
```diff
@@ -209,15 +306,53 @@ function isPrivateOrReservedIpv4([a, b, c]: [number, number, number, number]): b
...
+  return [...left, ...Array<number>(omitted).fill(0), ...right];
+}
+
 function isPrivateOrReservedIpv6(hostname: string): boolean {
-  if (!hostname.includes(":")) return false;
+  const hextets = parseIpv6Hextets(hostname);
+  if (!hextets) return false;
+
+  const first = hextets[0];
+  const unspecifiedOrIpv4Compatible = hextets.slice(0, 6).every((part) => part === 0);
```
[the remark this answers is not in the document]
Comment (SweetSophia) [reply openclaw-noosphere-memory/src/config.ts:336]: Verified against current head `65db0b82ebc81f9106a1fe9bbcacb69adc21cc70`: this is a false positive because `isBlockedInternalHost()` checks `isLoopbackHost(normalized)` first and returns `false` before `isPrivateOrReservedIpv6()` is called. The exact `https://[::1]:6578/` regression case passes in the focused suite and GitHub’s full `test` job. Narrowing the later reserved-range check as suggested would instead allow deprecated IPv4-compatible literals such as `::2`, so no code change is appropriate.

File: src/__tests__/memory/openclaw-plugin-auto-recall.test.ts [hunk 2]
```diff
@@ -1725,6 +1869,43 @@ describe("OpenClaw Noosphere corpus supplement", () => {
 });
 
 describe("OpenClaw Noosphere client", () => {
+  it("refuses redirects before request bodies can reach another origin", async () => {
+    const originalFetch = globalThis.fetch;
+    globalThis.fetch = async (_input, init) => {
+      assert.equal(init?.redirect, "error");
```
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/__tests__/memory/openclaw-plugin-auto-recall.test.ts]: Addressed in `65db0b82ebc81f9106a1fe9bbcacb69adc21cc70`: the test now starts two real loopback HTTP servers, returns a native 307 from the configured origin, and asserts the target receives zero requests, authorization headers, and body bytes. Sabotage verification removed `redirect: "error"`; native fetch followed the redirect and the test turned RED with `Missing expected rejection`. The guard was restored before commit.

**Comment by SweetSophia** [comment]
## Independent final review — PR #312

**Head reviewed:** `65db0b82ebc81f9106a1fe9bbcacb69adc21cc70` plus follow-up `3eb1b8174e0796cbf9728c8878cfc426eeeee212`.
**Method:** Fagan inspection + STRIDE on the OpenClaw credential path; live Node URL classification; focused plugin suite; bot-thread audit (refreshed after hosted reviewers finished).
**Verdict for stated OpenClaw scope:** the origin-pin claim holds. No blocking defect in this plugin. Residual same-class risk remains in sister plugins (out of this PR).

### Claim

An actor who can change `config.baseUrl` must not be able to send the env/secret API key to an untrusted origin. Remote destinations require `OPENCLAW_NOOSPHERE_TRUSTED_ORIGIN` (env-only HTTPS origin); `fetch` uses `redirect: "error"`.

### What was independently verified

- **Env-only pin.** `assertTrustedRemoteOrigin` in `openclaw-noosphere-memory/src/config.ts` reads `OPENCLAW_NOOSPHERE_TRUSTED_ORIGIN` then `NOOSPHERE_TRUSTED_ORIGIN` from `env` only. `NoosphereMemoryConfig` has no trust field; `openclaw.plugin.json` sets `additionalProperties: false`. Passing `trustedOrigins` in plugin config throws before any network call (`fetchCalls === 0`).
- **Single credentialed fetch.** `NoosphereMemoryClient.request` (`openclaw-noosphere-memory/src/client.ts`) is the only Authorization-bearing fetch. `redirect: "error"` is set after `...init`, so callers cannot override it. CLI `/api/health` fetch has no Authorization header.
- **Loopback vs internal.** `isLoopbackHost` no longer uses `startsWith("127.")`. `https://127.attacker.example` requires a pin. `::1` is exempt because `isBlockedInternalHost` short-circuits loopback before `isPrivateOrReservedIpv6`. Private IPv4/IPv6, mapped `::ffff:10.0.0.1`, expanded mapped form, `::2`, `::10.0.0.1`, ULA, link-local, site-local, discard-only, and documentation `2001:db8::` all fail closed.
- **URL canonicalization (live Node).** `http://2130706433`, `http://0177.0.0.1`, and `http://127.1` become `http://127.0.0.1` (loopback, intended). Default HTTPS port 443 matches a pin without `:443`. Unicode/punycode IDN origins match. Query, fragment, and embedded credentials throw. IPv6 zone IDs (`[fe80::1%25eth0]`) are rejected by `new URL` as invalid. Trailing-dot FQDN vs pin without the dot fails closed.
- **Native redirect test.** `refuses redirects before request bodies can reach another origin` uses two real `127.0.0.1` HTTP servers, a 307, and asserts the target received 0 requests, 0 Authorization headers, and 0 body bytes. Focused suite: **82/82 pass** after the follow-up tests (was 79/79 at `65db0b8`).
- **Fail-closed operator path.** Unpinned remote URLs throw `NoosphereConfigError` instead of silently falling back to localhost. Dist `config.js` matches src; `rejects unpinned remote OpenClaw endpoints in the built package` imports `dist/`.
- **CI at `65db0b8`:** lint, typecheck, test, package-check, CodeQL, Socket, docker, installer all green.

### Bot comments

| Bot | Finding | Verdict |
| --- | --- | --- |
| **Kilo** (thread on the redirect test) | Mocked `200` never exercised `redirect: "error"` | **Valid at `34c5db5`.** Closed in `65db0b8` with a native 307 pair. Kilo’s later “No Issues Found” is correct. The remaining “untested redirect” comments from other reviewers are stale. |
| **CodeRabbit** | Document `NOOSPHERE_TRUSTED_ORIGIN` fallback; say secret stores must inject into `process.env` | **Valid, then addressed** in `65db0b8` (README + `openclaw.plugin.json` help/description). Threads resolved. |
| **Kody** (`isPrivateOrReservedIpv6` blocks `::1`) | High-severity: `https://[::1]:6578/` would throw | **False positive on impact.** `isPrivateOrReservedIpv6("::1")` is `true` (first six hextets zero), but `isLoopbackHost` matches `::1` and `isBlockedInternalHost` returns `false` first. Live: `https://[::1]` succeeds even with a hostile pin in env. Author’s reply is correct: splitting the `::/96` check as suggested would allow deprecated IPv4-compatible literals such as `::2`. No code change. |
| **Sourcery** | Rate-limited; review guide only | No defect claims to judge. |

### Follow-up pushed in this review (`3eb1b81`)

In-scope hardening only:

1. `openclaw.plugin.json` `setup.providers.envVars` now lists `OPENCLAW_NOOSPHERE_TRUSTED_ORIGIN` and `NOOSPHERE_TRUSTED_ORIGIN`, matching the protected-env contract already documented in the README.
2. Regression tests for query/fragment/credentials, zone-ID URLs, IDN/punycode pin match, default port 443, and Node dword/octal/short IPv4 loopback canonicalization.

Sister-plugin ports were **not** mixed into this PR.

### Residual (do not block this PR)

These share the pre-fix threat model but are separate packages:

- **`opencode-noosphere-memory`:** `normalizeBaseUrl` still silently falls back to localhost on invalid URLs, does not block internal hosts, has no origin pin, and `fetch` defaults to `redirect: "follow"`. This is weaker than OpenClaw was before #312.
- **`kilocode-noosphere-memory`:** has internal-host blocking (still `startsWith("127.")`, so `127.attacker.example` is treated as loopback) but no pin and no `redirect: "error"`.
- **`hermes-noosphere-memory`:** solid `ipaddress` host checks, no pin; `urllib.request.urlopen` follows redirects (cross-host 307 can still send the body).

Recommended follow-up PR: port env-only origin pin + `redirect: "error"` (or a no-redirect opener) + fail-closed URL validation. Do not silently fall back to localhost.

Non-blocking notes for OpenClaw itself:

- Loopback remains pin-free on any port (`http://127.0.0.1:<any>`). Documented local-dev trust boundary, not a pin bypass.
- Pin is hostname-origin, not resolved IP. DNS rebinding of a *pinned* name is mitigated by mandatory remote HTTPS + TLS name checks. Worth one README sentence later, not a code change here.
- NAT64 (`64:ff9b::10.0.0.1`) and IPv4-translated (`::ffff:0:10.0.0.1`) embeddings of RFC1918 are pin-gated, not attacker-usable without the env pin.

### Security (STRIDE, OpenClaw path)

| Category | Result |
| --- | --- |
| Spoofing / Information disclosure | Mutable `baseUrl` cannot authorize itself; pin mismatch fails closed; credentials never appear in pin-error text. |
| Tampering | Plugin config cannot supply the pin. |
| Elevation | Remote HTTP rejected; internal/link-local/ULA rejected. |
| DoS | Fail-closed init; auto-recall still fails open for the agent while refusing to send the secret. |

**Merge for OpenClaw scope:** yes, after CI on `3eb1b81`. Do not treat this as authorization to leave Opencode/Kilo/Hermes unpinned.

File: src/__tests__/memory/openclaw-plugin-auto-recall.test.ts [hunk 3]
```diff
@@ -146,24 +147,70 @@ describe("resolveAutoRecallConfig", () => {
...
+      "https://noosphere.example.test?x=1",
+      "https://noosphere.example.test#frag",
+      "https://user:pass@noosphere.example.test",
+      "https://[fe80::1%25eth0]",
     ]) {
-      assert.equal(
-        resolveNoosphereMemoryConfig({ baseUrl }, {} as NodeJS.ProcessEnv).baseUrl,
-        DEFAULT_NOOSPHERE_BASE_URL,
+      assert.throws(
+        () => resolveNoosphereMemoryConfig({ baseUrl }, {} as NodeJS.ProcessEnv),
```
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/__tests__/memory/openclaw-plugin-auto-recall.test.ts]: The zone-ID impact claim is not valid on the supported Node URL path: `new URL("https://[fe80::1%25eth0]")` throws `TypeError: Invalid URL` before `isBlockedInternalHost()` runs, so there is no scoped-host value for the classifier to strip. Commit `a874ff8ea468a1560f9d46622eece436f2212129` makes the test contract explicit by moving this case into a dedicated assertion for `Noosphere base URL must be a valid HTTP or HTTPS URL.` No production parser change is appropriate.

**Comment by SweetSophia** [comment]
Follow-up triage of the independent review against current head `7ff9cb8b09bffbd22781893068d7c2dbf2efa4cf`:

- The central OpenClaw origin-pin verdict remains valid.
- Corrected one valid manifest issue in `a874ff8`: `setup.providers[].envVars` is provider-auth metadata, so trusted-origin URLs must not be listed as API-key candidates. A manifest regression now pins the credential-only list.
- Corrected valid test-placebo coverage in `a874ff8`: credential/query/fragment cases now run with a matching trusted-origin pin and assert the specific guard. Removing those guards turns the test RED.
- A fresh isolated pass found one remaining valid edge case: bare `?` and `#` delimiters were erased by WHATWG URL normalization because `.search`/`.hash` are empty. `7ff9cb8` now rejects raw delimiters; reverting to the old condition makes the regression RED (`Missing expected exception`). Percent-encoded `%3F`/`%23` path data remains accepted.
- The zone-ID bot claim remains invalid for this runtime: Node rejects `https://[fe80::1%25eth0]` as `Invalid URL`. The test now asserts that exact parser path instead of implying host-classifier coverage.

Verification: focused suite 85/85, typecheck, lint (two unchanged unrelated warnings), package policy, package build/typecheck, deterministic committed `dist`, pack dry-run, and current-head GitHub CI all green. Second isolated review: `PASS`, zero findings.

**Comment by SweetSophia** [comment]
## Independent re-review — PR #312 @ `7ff9cb8`

Delta only vs prior review at `65db0b8` + `3eb1b81` ([issuecomment-5477225041](https://github.com/SweetSophia/noosphere/pull/312#issuecomment-5477225041)). Not a restatement of that review.

**New commits:** `a874ff8` (keep origin pins out of auth metadata), `7ff9cb8` (reject empty URL delimiters).
**Evidence:** live Node classification against committed `dist/config.js`, focused plugin suite **85/85**, GitHub CI green at this head, review threads refreshed after hosted bots finished (5 threads, all resolved).

### Verdict

No blocking defect at `7ff9cb8`. The OpenClaw origin-pin claim still holds. Merge for OpenClaw scope.

### `a874ff8` — envVars revert

**My prior recommendation to list `OPENCLAW_NOOSPHERE_TRUSTED_ORIGIN` / `NOOSPHERE_TRUSTED_ORIGIN` in `setup.providers[].envVars` was wrong.**

That array is provider-auth metadata (API-key candidates), not a general protected-env catalog. Putting the pin there would advertise a non-secret origin URL as an auth credential. Current list is correct:

```json
"envVars": ["OPENCLAW_NOOSPHERE_API_KEY", "NOOSPHERE_API_KEY"]
```

`keeps trusted-origin variables out of OpenClaw provider auth metadata` now locks that contract. Pin still comes only from `process.env` via `assertTrustedRemoteOrigin` (`openclaw-noosphere-memory/src/config.ts`).

Credential / query / fragment cases now run **with a matching pin** and assert the specific guard. That is real coverage, not a placebo that fails later at the pin check.

### `7ff9cb8` — empty `?` / `#`

Valid invariant hole, **not a pin bypass**.

WHATWG `new URL("https://host?")` yields `search === ""` (falsy), so `url.search || url.hash` accepted a URL the policy said to reject. Origin was still `https://host`; credentials would not have left the pinned origin. Switching `normalizeBaseUrl` to `trimmed.includes("?") || trimmed.includes("#")` matches the written policy.

Live against `dist/config.js` with a matching pin:

| Input | Result |
| --- | --- |
| `https://noosphere.example.test?` / `#` / `/?` / `/#` | rejected (query/fragment) |
| `https://noosphere.example.test/foo%3Fbar` and `%23` | accepted (path data) |
| `http://127.0.0.1:6578?` | rejected even on loopback |

`normalizeTrustedOrigin` still uses `!!url.search || !!url.hash`. Live: pin values `https://host?` and `https://host#` still authorize `https://host` because `.origin` is unchanged. Operator-controlled env, same origin — not a bypass. Do not “fix” unless you want pin syntax as strict as baseUrl.

### New Kody thread (zone ID, 10:53)

**False positive.** Same class as the `::1` finding.

`new URL("https://[fe80::1%25eth0]")` throws `TypeError: Invalid URL` before `isBlockedInternalHost`. There is no scoped hostname for the classifier to strip. Dedicated test now asserts `must be a valid HTTP or HTTPS URL.` Author reply on the thread is correct. No production parser change.

Kilo’s later “No Issues Found” on the delimiter commit is correct (the `/[?#]/` rewrite is style, not a defect).

### Unchanged residuals (still not this PR)

Opencode fail-opens to localhost with no pin and follows redirects. Kilo still `startsWith("127.")` and has no pin. Hermes `urlopen` follows redirects. Same threat class, separate packages.

No further code change from this pass.
